src/celeritas/optical/ScintillationPreGenerator.hh

```diff
@@ -0,0 +1,143 @@
+#include "celeritas/random/distribution/PoissonDistribution.hh"
+#include "celeritas/track/SimTrackView.hh"
+
+#include "OpticalDistributionData.hh"
+#include "OpticalPropertyData.hh"
+#include "ScintillationData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the number of Scintillation photons to be generated by
+ * \c ScintillationGenerator and populate \c OpticalDistributionData values
+ * using Param and State data.
+ */
+class ScintillationPreGenerator
+{
+  public:
+    // Placeholder for data that is not available through Views
+    // TODO: Merge Cerenkov and Scintillation pre-gen data into one struct
+    struct OpticalPreGenStepData
+    {
+        real_type time{};  //!< Pre-step time
+        units::MevEnergy energy_dep;  //!< Step energy deposition
+        EnumArray<StepPoint, OpticalStepData> points;  //!< Pre- and post-steps
+
+        //! Check whether the data are assigned
+        explicit CELER_FUNCTION operator bool() const
+        {
+            return energy_dep > zero_quantity()
+                   && points[StepPoint::pre].speed > zero_quantity();
+        }
+    };
+
+    // Construct with optical properties, scintillation, and step data
+    inline CELER_FUNCTION
+    ScintillationPreGenerator(ParticleTrackView const& particle_view,
+                              SimTrackView const& sim_view,
+                              OpticalMaterialId material,
+                              NativeCRef<ScintillationData> const& shared,
+                              OpticalPreGenStepData const& step_data);
+
+    // Populate an optical distribution data for the Scintillation Generator
+    template<class Generator>
+    inline CELER_FUNCTION OpticalDistributionData operator()(Generator& rng);
+
+  private:
+    units::ElementaryCharge charge_;
+    real_type step_len_;
+    OpticalMaterialId mat_id_;
+    NativeCRef<ScintillationData> const& shared_;
+    OpticalPreGenStepData step_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with input parameters.
+ */
+inline CELER_FUNCTION ScintillationPreGenerator::ScintillationPreGenerator(
+    ParticleTrackView const& particle_view,
+    SimTrackView const& sim_view,
+    OpticalMaterialId material,
+    NativeCRef<ScintillationData> const& shared,
+    OpticalPreGenStepData const& step_data)
+    : charge_(particle_view.charge())
+    , step_len_(sim_view.step_length())
+    , mat_id_(material)
+    , shared_(shared)
+    , step_(step_data)
+{
+    CELER_EXPECT(step_len_ > 0);
+    CELER_EXPECT(mat_id_);
+    CELER_EXPECT(shared_);
+    CELER_EXPECT(step_);
+
+    if (shared_.scintillation_by_particle())
+    {
+        // TODO: implement sampling for particles
+        CELER_NOT_IMPLEMENTED("scintillation by particle type");
```

Comment:
the "not implemented" should be already handled by the earlier input validation, and this function is meant mainly for host code. Maybe `CELER_ASSERT_UNREACHABLE()` with a comment...